contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,193 @@
+# topology aware routing of services
+
+Author: @m1093782566
+
+## Objective
+
+Provide a generic way to provide the local-only service route, say "topology aware routing of service". Local means "same topology level", e.g. same node, same rack, same failure zone, same failure region, same cloud provider etc. To clarify, this proposal is intended to resolve the user case that user always want to visit a local service's backend even though for some reason local service's backend is not available.
+
+## GOAL
+
+A generic way to support various topology levels, regarless whatever the topology key is.
+
+## Non-goal
+
+Scheduler spreading to implement this sort of topology guarantee.
+
+## Use cases
+
+* Logging agents such as fluentd. Deploy fluentd as DaemonSet and applications only need to communicate with the fluentd in the same node.
+* For a sharded service that keeps per-node local information in each shard.
+* Authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy).
+* In container identity wg, being able to give daemonset pods a unique identity per host is on the 2018 plan, and ensuring local pods can communicate to local node services securely is a key goal there. -- from @smarterclayton
+* Regional data costs in multi-AZ setup - for instance, in AWS, with a multi-AZ setup, half of the traffic will switch AZ, incurring regional data Transfer costs, whereas if something was local, it wouldn't hit the network.
+* Users want a priority-based topology aware routing service. For example, want to visit a "same-zone" service,  of course, a "same-host" backend is more preferred.
+
```

Comment:
Performance benefit (node local/rack local) is lower latency/higher bandwidth?

Reply:
Yes, it's another good user case. It's added now, thanks!